### PR TITLE
Issue/7 json complete mode

### PR DESCRIPTION
**Pull request by bogdanadrianmarc**
This PR addresses issue epimorphics/ds-api-ruby#7 by adding support for the `complete` JSON mode. The issue is that PPD uses JSON compact mode and UKHPI uses JSON complete mode, and currently the latter is not supported; this PR adds that. The repo with the new SAPINT API for LR is here: https://github.com/epimorphics/lr-data-api (I'm using the master branch). From my tests with the LR apps nothing broke by adding this and UKHPI actually works now, but please run this to double check.

@ijdickinson's suggestions from last PR did not carry over, but I will address them here in a moment

### CODE REVIEW

**Review by ijdickinson — CHANGES_REQUESTED** [review]
On the whole I approve of these changes, :+1: 

I'm still not happy about the design of `json_mode_complete` though. I've put in the comment some suggestions for changing the implementation of that method to something that is cleaner and more maintainable. There is another approach though, which would be to do some investigation to see how many changes would need to be made to UKHPI for it to use compact mode rather than complete mode. I have no idea off the top of my head, but it may be that the amount of work you need to do to robustly support both compact and complete modes is more than just adapting the app to use the same mode as PPD, and that would make the code more maintainable in the long run. Worth a short investigation, I would think.

File: spec/data_services_api/dataset_spec.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@
 describe 'DataServiceApi::Dataset' do
   before do
     VCR.insert_cassette name, record: :new_episodes
-    @dataset = DataServicesApi::Service.new(url: 'http://localhost:8080').dataset('ukhpi')
+    @dataset = DataServicesApi::Service.new(url: 'http://localhost:8888').dataset('ukhpi')
```
Comment (ijdickinson) [inline spec/data_services_api/dataset_spec.rb:8]: This should be configurable, or there should be instructions in the readme as to what to set up on `localhost:8888`

Comment (bogdanadrianmarc) [reply spec/data_services_api/dataset_spec.rb:8]: What should the readme contain? A link to the new SAPI API for LR?

Comment (ijdickinson) [reply spec/data_services_api/dataset_spec.rb:8]: The key test is whether someone who doesn't have the benefit of already having worked on the project (or did, but six months ago and has now forgotten!) can recreate a working setup using just these instructions. As it stands, I think the answer would be "no". They would have to ask someone how to do some of the steps.

A general principle for these kinds of situation is to not hard-code the URL into the test, but instead put it into an environment variable. Then, you can say in the readme what variables to set in the environment, what the default values are, and, if needed, how to set up your dev system so that the tests will work

File: lib/data_services_api/dsapi_response_converter.rb [hunk 2]
```diff
@@ -23,14 +24,33 @@ def to_dsapi_item(sapint_item)
...
-        ["#{dataset_name}:#{sapint_key}#{key == '@id' ? '' : key.capitalize}", value]
+        ["#{@dataset_name}:#{sapint_key}#{key == '@id' ? '' : key.capitalize}", value]
       end.to_h
     end
+
+    def json_mode_complete(sapint_key, sapint_value)
+      case sapint_key
+      when '@id'
+        return { sapint_key => sapint_value }
+      when 'refMonth'
```
[the remark this answers is not in the document]
Comment (bogdanadrianmarc) [reply lib/data_services_api/dsapi_response_converter.rb:46]: I agree with the suggestions, I wouldn't have gone this route if I had more information about the fields. The problem is that, SAPINT returns:
Screenshot 2021-11-01 at 14 50 40
Screenshot 2021-11-01 at 14 50 51

for `refMonth` and `refPeriodStart`, which are essentially 2 strings that cannot be differentiated by anything else other than field name. There's no indication here that these fields should be in this format (which is returned by the original DSAPI and is expected by UKHPI):
Screenshot 2021-11-01 at 14 53 37
Screenshot 2021-11-01 at 14 53 23

This information is not found in the original query either, so I'm not able to figure out that these 2 fields should be treated differently from the query. The best I can do is provide a configuration type Hash that I would iterate over and format the required fields accordingly

[the remark this answers is not in the document]
Comment (bogdanadrianmarc) [reply lib/data_services_api/dsapi_response_converter.rb:46]: Making a note here that we discussed this on Slack and created a new issue to come back to later on, here: https://github.com/epimorphics/ds-api-ruby/issues/9

File: README.md [hunk 3]
```diff
@@ -6,10 +6,26 @@ This gem provides a Ruby API for the [data services API](https://github.com/epim
...
 And then execute:
 
-    $ bundle
+```sh
+bundle
+```
 
-Also, an API URL needs to be provided for the Service class in order for the gem to work
\ No newline at end of file
+Also, an API URL needs to be provided for the Service class in order for the gem
```
Comment (mairead) [inline README.md:19]: An explanation of why here might be useful?

Comment (bogdanadrianmarc) [reply README.md:19]: There's no fallback default value in the gem for the API URL, so if not specified the gem won't work. I'm gonna rephrase this to something like "you need to set <this> as an environment variable" so it's more clear to the gem user
